File: api-refund-not-updating-order-line/python/fix_api_refund_quantity.py
"""Find and repair PrestaShop order lines whose refunded quantity is stale
after a credit slip was created through the webservice.

POST /api/order_slip only inserts rows into order_slip and order_slip_detail. It
never runs the back office refund logic in OrderSlip::create() or
AdminOrdersController, which is what actually recalculates and writes
order_detail.product_quantity_refunded, the refund totals, and the related stock
movement. So a credit slip can exist while the order line still reports its old
refunded quantity.

This script sums product_quantity from every order_slip_detail row per
id_order_detail to get the expected refunded quantity, compares it against the
stored product_quantity_refunded, and only writes the corrected value when
DRY_RUN is explicitly false. A negative delta (stored higher than expected) is
always flagged for a human, never auto-corrected.

Guide: https://www.allanninal.dev/prestashop/api-refund-not-updating-order-line/

Run on demand for a suspected order id, or on a schedule across recent orders.
Safe to run again and again.
"""
import os
import logging
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("fix_api_refund_quantity")

PRESTASHOP_URL = os.environ.get("PRESTASHOP_URL", "https://demo.example.com").rstrip("/")
PRESTASHOP_WS_KEY = os.environ.get("PRESTASHOP_WS_KEY", "WSKEYDUMMY")
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"
ORDER_IDS = os.environ.get("ORDER_IDS", "1,2,3")
AUTH = (PRESTASHOP_WS_KEY, "")


def compute_refund_delta(stored_refunded_qty, order_slip_quantities):
    """Pure decision logic, no I/O.

    Sums order_slip_quantities to get the expected refunded quantity, and
    compares it against stored_refunded_qty. needs_repair means the API-created
    credit slips claim more refunded units than the order line shows, the
    exact symptom this script exists to fix. needs_review means the stored
    value is already higher than the credit slips justify, which is left for
    a human rather than corrected automatically.
    """
    expected = sum(order_slip_quantities)
    delta = expected - stored_refunded_qty
    return {
        "expected": expected,
        "stored": stored_refunded_qty,
        "delta": delta,
        "needs_repair": expected > stored_refunded_qty,
        "needs_review": expected < stored_refunded_qty,
    }


def api_get(path, params=None):
    params = dict(params or {})
    params["output_format"] = "JSON"
    r = requests.get(f"{PRESTASHOP_URL}/api/{path}", params=params, auth=AUTH, timeout=30)
    r.raise_for_status()
    return r.json()


def order_slips_for(id_order):
    data = api_get("order_slip", params={"filter[id_order]": id_order, "display": "full"})
    return data.get("order_slips") or []


def order_detail(id_order_detail):
    data = api_get(f"order_details/{id_order_detail}", params={"display": "full"})
    return data["order_detail"]


def order_history_states(id_order):
    data = api_get("order_histories", params={"filter[id_order]": id_order, "display": "full"})
    return [h.get("id_order_state") for h in (data.get("order_histories") or [])]


def refund_quantities_by_line(order_slips):
    """Group order_slip_detail rows by id_order_detail and collect their quantities."""
    by_line = {}
    for slip in order_slips:
        details = (slip.get("associations", {}) or {}).get("order_slip_detail") or slip.get("order_slip_detail") or []
        for row in details:
            id_order_detail = row["id_order_detail"]
            by_line.setdefault(id_order_detail, []).append(int(row["product_quantity"]))
    return by_line


def apply_expected_refund(id_order_detail, expected_qty):
    full = api_get(f"order_details/{id_order_detail}")["order_detail"]
    unit_price_tax_excl = float(full.get("unit_price_tax_excl", 0) or 0)
    unit_price_tax_incl = float(full.get("unit_price_tax_incl", 0) or 0)
    full["product_quantity_refunded"] = expected_qty
    if "total_refunded_tax_excl" in full:
        full["total_refunded_tax_excl"] = f"{expected_qty * unit_price_tax_excl:.6f}"
    if "total_refunded_tax_incl" in full:
        full["total_refunded_tax_incl"] = f"{expected_qty * unit_price_tax_incl:.6f}"
    r = requests.put(
        f"{PRESTASHOP_URL}/api/order_details/{id_order_detail}",
        params={"output_format": "JSON"},
        json={"order_detail": full},
        auth=AUTH,
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def run():
    checked = 0
    repaired = 0
    flagged_for_review = 0
    for id_order in [s.strip() for s in ORDER_IDS.split(",") if s.strip()]:
        slips = order_slips_for(id_order)
        if not slips:
            continue
        by_line = refund_quantities_by_line(slips)
        history_states = order_history_states(id_order)
        for id_order_detail, quantities in by_line.items():
            detail = order_detail(id_order_detail)
            stored = int(detail.get("product_quantity_refunded", 0) or 0)
            result = compute_refund_delta(stored, quantities)
            checked += 1
            if result["delta"] == 0:
                continue
            if result["needs_review"]:
                flagged_for_review += 1
                log.warning(
                    "Needs human review. id_order=%s id_order_detail=%s stored=%d expected=%d delta=%d",
                    id_order, id_order_detail, result["stored"], result["expected"], result["delta"],
                )
                continue
            if not history_states:
                flagged_for_review += 1
                log.warning(
                    "Skipping repair, no order_histories rows found. id_order=%s id_order_detail=%s",
                    id_order, id_order_detail,
                )
                continue
            log.info(
                "Refund quantity stale. id_order=%s id_order_detail=%s stored=%d expected=%d %s",
                id_order, id_order_detail, result["stored"], result["expected"],
                "would repair" if DRY_RUN else "repairing",
            )
            if not DRY_RUN:
                apply_expected_refund(id_order_detail, result["expected"])
                verify = order_detail(id_order_detail)
                log.info(
                    "Verified. id_order_detail=%s product_quantity_refunded=%s",
                    id_order_detail, verify.get("product_quantity_refunded"),
                )
            repaired += 1
    log.info(
        "Done. %d line(s) checked, %d repaired, %d flagged for review. DRY_RUN=%s",
        checked, repaired, flagged_for_review, DRY_RUN,
    )


if __name__ == "__main__":
    run()

File: api-refund-not-updating-order-line/node/fix-api-refund-quantity.js
/**
 * Find and repair PrestaShop order lines whose refunded quantity is stale
 * after a credit slip was created through the webservice.
 *
 * POST /api/order_slip only inserts rows into order_slip and order_slip_detail. It
 * never runs the back office refund logic in OrderSlip::create() or
 * AdminOrdersController, which is what actually recalculates and writes
 * order_detail.product_quantity_refunded, the refund totals, and the related stock
 * movement. So a credit slip can exist while the order line still reports its old
 * refunded quantity.
 *
 * This script sums product_quantity from every order_slip_detail row per
 * id_order_detail to get the expected refunded quantity, compares it against the
 * stored product_quantity_refunded, and only writes the corrected value when
 * DRY_RUN is explicitly false. A negative delta (stored higher than expected) is
 * always flagged for a human, never auto-corrected.
 *
 * Guide: https://www.allanninal.dev/prestashop/api-refund-not-updating-order-line/
 */
import { pathToFileURL } from "node:url";

const PRESTASHOP_URL = (process.env.PRESTASHOP_URL || "https://demo.example.com").replace(/\/+$/, "");
const PRESTASHOP_WS_KEY = process.env.PRESTASHOP_WS_KEY || "WSKEYDUMMY";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";
const ORDER_IDS = process.env.ORDER_IDS || "1,2,3";

function basicAuthHeader() {
  return "Basic " + Buffer.from(`${PRESTASHOP_WS_KEY}:`).toString("base64");
}

/**
 * Pure decision logic, no I/O.
 *
 * Sums orderSlipQuantities to get the expected refunded quantity, and compares
 * it against storedRefundedQty. needsRepair means the API-created credit slips
 * claim more refunded units than the order line shows. needsReview means the
 * stored value is already higher than the credit slips justify, which is left
 * for a human rather than corrected automatically.
 */
export function computeRefundDelta(storedRefundedQty, orderSlipQuantities) {
  const expected = orderSlipQuantities.reduce((a, b) => a + b, 0);
  const delta = expected - storedRefundedQty;
  return {
    expected,
    stored: storedRefundedQty,
    delta,
    needs_repair: expected > storedRefundedQty,
    needs_review: expected < storedRefundedQty,
  };
}

async function apiGet(path, params = {}) {
  const url = new URL(`${PRESTASHOP_URL}/api/${path}`);
  url.searchParams.set("output_format", "JSON");
  for (const [key, value] of Object.entries(params)) url.searchParams.set(key, value);
  const res = await fetch(url, { headers: { Authorization: basicAuthHeader() } });
  if (!res.ok) throw new Error(`PrestaShop ${res.status} on GET ${path}`);
  return res.json();
}

async function orderSlipsFor(idOrder) {
  const data = await apiGet("order_slip", { "filter[id_order]": idOrder, display: "full" });
  return data.order_slips || [];
}

async function orderDetail(idOrderDetail) {
  const data = await apiGet(`order_details/${idOrderDetail}`, { display: "full" });
  return data.order_detail;
}

async function orderHistoryStates(idOrder) {
  const data = await apiGet("order_histories", { "filter[id_order]": idOrder, display: "full" });
  return (data.order_histories || []).map((h) => h.id_order_state);
}

function refundQuantitiesByLine(orderSlips) {
  const byLine = new Map();
  for (const slip of orderSlips) {
    const details = (slip.associations && slip.associations.order_slip_detail) || slip.order_slip_detail || [];
    for (const row of details) {
      const idOrderDetail = row.id_order_detail;
      if (!byLine.has(idOrderDetail)) byLine.set(idOrderDetail, []);
      byLine.get(idOrderDetail).push(Number(row.product_quantity));
    }
  }
  return byLine;
}

async function applyExpectedRefund(idOrderDetail, expectedQty) {
  const full = (await apiGet(`order_details/${idOrderDetail}`)).order_detail;
  const unitPriceTaxExcl = Number(full.unit_price_tax_excl || 0);
  const unitPriceTaxIncl = Number(full.unit_price_tax_incl || 0);
  full.product_quantity_refunded = expectedQty;
  if ("total_refunded_tax_excl" in full) full.total_refunded_tax_excl = (expectedQty * unitPriceTaxExcl).toFixed(6);
  if ("total_refunded_tax_incl" in full) full.total_refunded_tax_incl = (expectedQty * unitPriceTaxIncl).toFixed(6);
  const url = new URL(`${PRESTASHOP_URL}/api/order_details/${idOrderDetail}`);
  url.searchParams.set("output_format", "JSON");
  const res = await fetch(url, {
    method: "PUT",
    headers: { Authorization: basicAuthHeader(), "Content-Type": "application/json" },
    body: JSON.stringify({ order_detail: full }),
  });
  if (!res.ok) throw new Error(`PrestaShop ${res.status} on PUT order_details/${idOrderDetail}`);
  return res.json();
}

export async function run() {
  let checked = 0;
  let repaired = 0;
  let flaggedForReview = 0;
  const orderIds = ORDER_IDS.split(",").map((s) => s.trim()).filter(Boolean);
  for (const idOrder of orderIds) {
    const slips = await orderSlipsFor(idOrder);
    if (!slips.length) continue;
    const byLine = refundQuantitiesByLine(slips);
    const historyStates = await orderHistoryStates(idOrder);
    for (const [idOrderDetail, quantities] of byLine.entries()) {
      const detail = await orderDetail(idOrderDetail);
      const stored = Number(detail.product_quantity_refunded || 0);
      const result = computeRefundDelta(stored, quantities);
      checked++;
      if (result.delta === 0) continue;
      if (result.needs_review) {
        flaggedForReview++;
        console.warn(
          `Needs human review. id_order=${idOrder} id_order_detail=${idOrderDetail} ` +
            `stored=${result.stored} expected=${result.expected} delta=${result.delta}`
        );
        continue;
      }
      if (!historyStates.length) {
        flaggedForReview++;
        console.warn(`Skipping repair, no order_histories rows found. id_order=${idOrder} id_order_detail=${idOrderDetail}`);
        continue;
      }
      console.log(
        `Refund quantity stale. id_order=${idOrder} id_order_detail=${idOrderDetail} ` +
          `stored=${result.stored} expected=${result.expected} ${DRY_RUN ? "would repair" : "repairing"}`
      );
      if (!DRY_RUN) {
        await applyExpectedRefund(idOrderDetail, result.expected);
        const verify = await orderDetail(idOrderDetail);
        console.log(`Verified. id_order_detail=${idOrderDetail} product_quantity_refunded=${verify.product_quantity_refunded}`);
      }
      repaired++;
    }
  }
  console.log(`Done. ${checked} line(s) checked, ${repaired} repaired, ${flaggedForReview} flagged for review. DRY_RUN=${DRY_RUN}`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
